lenskit-slim/src/main/java/org/lenskit/slim/AbstractLinearRegression.java

```diff
@@ -0,0 +1,138 @@
+
+    @Inject
+    public AbstractLinearRegression(SlimUpdateParameters updateParameters) {
+        this.updateParameters = updateParameters;
+    }
+
+    /**
+     * compute prediction of single observation
+     * @param observation test data
+     * @param weight weight
+     * @return prediction
+     */
+    public double computePrediction(Long2DoubleMap observation, Long2DoubleMap weight) {
+        return Vectors.dotProduct(weight, observation);
+    }
+
+    /**
+     * compute residual vector
+     * @param labels label vector
+     * @param dataMatrix column-wise data matrix
+     * @param weights weights vector
+     * @return residual vector
+     */
+    public Long2DoubleMap computeResiduals(Long2DoubleMap labels, Map<Long, Long2DoubleMap> dataMatrix, Long2DoubleMap weights) {
+        Long2DoubleMap predictions = predict(dataMatrix, weights);
+        return addVectors(labels, Vectors.multiplyScalar(predictions,-1.0));
+    }
+
+    /**
+     * learning process
+     *
+     * @param labels label vector
+     * @param trainingDataMatrix observations matrix row: user ratings for different items, column: item ratings of different users
+     * @return a trained weight vector
+     */
+    public abstract Long2DoubleMap fit(Long2DoubleMap labels, Map<Long, Long2DoubleMap> trainingDataMatrix);
+
+    /**
+     * learning process passed in pre-computed inner-products to speed up learning iterations
+     *
+     * @param labels label vector
+     * @param trainingDataMatrix Map of Item IDs to item rating vectors.
+     * @param covM Map of Item IDs to item-item inner-products vectors
+     * @param item item ID of label vector {@code labels}
+     * @return a trained weight vector
+     */
+    public abstract Long2DoubleMap fit(Long2DoubleMap labels, Map<Long, Long2DoubleMap> trainingDataMatrix, Map<Long, Long2DoubleMap> covM, long item);
+
+
+    /**
+     * optimized computation to update residual after one element of weight vector changed
+     * @param r original residual vector
+     * @param column one column of training matrix whose index equal to index of weight that is just updated
+     * @param weightToUpdate one element of weight vector before update
+     * @param weightUpdated one element of weight vector after update
+     * @return
+     */
+    public Long2DoubleMap updateResiduals(Long2DoubleMap r, Long2DoubleMap column, double weightToUpdate, double weightUpdated) {
+        Long2DoubleMap residualsIncrement = Vectors.multiplyScalar(column, (weightToUpdate - weightUpdated));
+        Long2DoubleMap residulsUpdated = addVectors(r, residualsIncrement);
+        return residulsUpdated;
+    }
+
+    /**
+     * compute predictions given test data matrix and learned weights
+     * @param testData column-wise data matrix
+     * @param weights weight
+     * @return prediction vector
+     */
+    public Long2DoubleMap predict(Map<Long, Long2DoubleMap> testData, Long2DoubleMap weights) {
```

Comment:
We can have `Long2ObjectMap<Long2DoubleMap>`.

Although really, given how much they're used here and elsewhere, we should probably have a `Long2Long2DoubleMap` or `LLDSparseMatrix` or something like that, that encapsulates the concept. Not that you have to build such a thing for this PR, it just seems like it'd be useful here.